src/Compiler/Checking/ConstraintSolver.fs

```diff
@@ -2942,10 +2942,16 @@ and ResolveOverloading
     let candidates = calledMethGroup |> List.filter (fun cmeth -> cmeth.IsCandidate(m, ad))
 
     let calledMethOpt, errors, calledMethTrace = 
-
         match calledMethGroup, candidates with 
-        | _, [calledMeth] when not isOpConversion -> 
-            Some calledMeth, CompleteD, NoTrace
+        | _, [calledMeth] when not isOpConversion ->
+            // See what candidates we have based on static/virtual/abstract
+            // OK: static virtual TResult operator checked i.e. IAdditionOperators.op_CheckedAddition
+            // Error: static abstract TResult operator i.e. IAdditionOperators.op_Addition
+            match calledMeth.Method with
+            | ILMeth(ilMethInfo= ilMethInfo) when ilMethInfo.IsStatic && ilMethInfo.IsVirtual && methodName.Contains("op_") && ilMethInfo.IsAbstract ->
```

Comment:
```suggestion
            | ILMeth(ilMethInfo= ilMethInfo) when ilMethInfo.IsStatic && ilMethInfo.IsVirtual && ilMethInfo.IsAbstract ->
```

Why special-case for `op_` members only, should this not apply to all methods generally the same?

Reply:
![Screenshot 2024-04-11 at 13 40 38](https://github.com/dotnet/fsharp/assets/31915729/3e312ac1-ab06-4cf9-a5b0-75f8dffbe4f4)

@T-Gro 'T.Parse should work through multiple levels of inheritance, since it will know specific implementation at runtime. ISpanParsable<'T>.Parse, if it's non-virtual, shouldn't on other hand.

Reply:
cc @vzarytovskii  Pointed out this in a previous PR

Reply:
> 'T.Parse should work through multiple levels of inheritance, since it will know specific implementation at runtime

Doesn't that apply to `'T.op_Addition` (or `'T.(+)`) as well, though?

As far as I can tell, the behavior depends on abstractness, not on operatorness:

```fsharp
open System
open System.Numerics

module ConstrainedCall =
    let ``'T.op_Addition``<'T & #IAdditionOperators<'T, 'T, 'T>>        x y    = 'T.op_Addition        (x, y)
    let ``'T.(+)``<'T & #IAdditionOperators<'T, 'T, 'T>>                x y    = 'T.(+)                (x, y)
    let ``'T.op_CheckedAddition``<'T & #IAdditionOperators<'T, 'T, 'T>> x y    = 'T.op_CheckedAddition (x, y)
    let ``'T.Parse``<'T & #IParsable<'T>>                               x      = 'T.Parse              (x, null)

module InterfaceCall =
    let ``IAdditionOperators.op_Addition``                              x y    = IAdditionOperators.op_Addition        (x, y)
    let ``IAdditionOperators.(+)``                                      x y    = IAdditionOperators.(+)                (x, y)
    let ``IAdditionOperators.op_CheckedAddition``                       x y    = IAdditionOperators.op_CheckedAddition (x, y)
    let ``IParsable.Parse``<'T & #IParsable<'T>>                        x : 'T = IParsable.Parse                       (x, null)

open ConstrainedCall

printfn $"{``'T.op_Addition`` 1 2}"                         // Prints 3.
printfn $"{``'T.(+)`` 1 2}"                                 // Prints 3.
printfn $"{``'T.op_CheckedAddition`` 1 2}"                  // Prints 3.
printfn $"""{``'T.Parse``<int> "3"}"""                      // Prints 3.

open InterfaceCall

printfn $"{``IAdditionOperators.op_Addition`` 1 2}"         // System.BadImageFormatException: Bad IL format.
printfn $"{``IAdditionOperators.(+)`` 1 2}"                 // System.BadImageFormatException: Bad IL format.
printfn $"{``IAdditionOperators.op_CheckedAddition`` 1 2}"  // Prints 3.
printfn $"""{``IParsable.Parse``<int> "3"}"""               // System.BadImageFormatException: Bad IL format.
```

Am I missing something?

Reply:
> Doesn't that apply to 'T.op_Addition (or 'T.(+)) as well, though?

@brianrourkeboll That is what I thought but I was told that `T.Parse` (constrained call) should allowed so that is why had to check for `op_`.  so only `static virtual` would be allowed.

Read the https://github.com/fsharp/fslang-design/blob/main/FSharp-7.0/FS-1124-interfaces-with-static-abstract-members.md and It does not clarify this scenario.

Reply:
> That is what I thought but I was told that `T.Parse` (constrained call) should be allowed so that is why had to check for `op_`. so only `static virtual` would be allowed.

Hmm, I still don't think I understand — unless you're talking about the bug being addressed in #17013?

If you look at my snippet above, the current behavior is that the constrained calls (like `'T.Parse`) all compile and run — and _should_ both compile and run — regardless of whether the static method guaranteed by the constraint is `abstract`, `virtual`, or an operator.

The direct calls on the interfaces also all currently compile regardless of whether the method guaranteed by the constraint is `abstract`, `virtual`, or an operator.

The only kind of invocation that fails at runtime with a bad image format is static `abstract` methods invoked directly on interfaces — but this is the case regardless of whether they are operators.

```fsharp
open System
open System.Numerics

// Is an operator, is abstract, fails at runtime, should not compile.
let ``IAdditionOperators.op_Addition`` x y = IAdditionOperators.op_Addition (x, y)
// Is an operator, is abstract, fails at runtime, should not compile.
let ``IAdditionOperators.(+)`` x y = IAdditionOperators.(+) (x, y)
// Is an operator, is virtual, currently runs fine.
let ``IAdditionOperators.op_CheckedAddition`` x y = IAdditionOperators.op_CheckedAddition (x, y)
// Is not an operator, is abstract, fails at runtime, should not compile.
let ``IParsable.Parse``<'T & #IParsable<'T>> x : 'T = IParsable.Parse (x, null)

printfn $"{``IAdditionOperators.op_Addition`` 1 2}"         // System.BadImageFormatException: Bad IL format.
printfn $"{``IAdditionOperators.(+)`` 1 2}"                 // System.BadImageFormatException: Bad IL format.
printfn $"{``IAdditionOperators.op_CheckedAddition`` 1 2}"  // Prints 3.
printfn $"""{``IParsable.Parse``<int> "3"}"""               // System.BadImageFormatException: Bad IL format.
```

C# does not allow calling static `abstract` or `virtual` methods on interfaces at all:

```csx
> IParsable<int>.Parse("3", null)
(1,1): error CS8926: A static virtual or abstract interface member can be accessed only on a type parameter.
```

```csx
> ISpanParsable<int>.Parse("3", null)
(1,1): error CS8926: A static virtual or abstract interface member can be accessed only on a type parameter.
```

It also just does not allow directly invoking operators in general, regardless of whether they are on interfaces:

```csx
> IAdditionOperators<int, int, int>.op_Addition(1, 2)
(1,35): error CS0571: 'IAdditionOperators<int, int, int>.operator +(int, int)': cannot explicitly call operator or accessor
```

```csx
> IAdditionOperators<int, int, int>.op_CheckedAddition(1, 2)
(1,35): error CS0571: 'IAdditionOperators<int, int, int>.operator checked +(int, int)': cannot explicitly call operator or accessor
```

```csx
> public class C { public static int operator +(C x, C y) => 3; }
. C.op_Addition(new(), new())
(2,3): error CS0571: 'C.operator +(C, C)': cannot explicitly call operator or accessor
```

F# of course _does_ allow this in general, both with `(…)` syntax and explicitly uttering the compiled `op_…` name. The only reason operators seem to stand out here is because operators are always static, and operators are one of the main reasons for static methods in interfaces in the first place.

It's too late for F# to follow C# and disallow direct invocations of `virtual` methods on interfaces altogether, since that _does_ currently compile and (sometimes) run — regardless of whether the method is an operator.

But I think the compile-time error that we're adding here, for when a static `abstract` method is called directly on an interface, needs to apply regardless of whether the `abstract` method is an operator.

Reply:
Exactly, would apply it generally without the `op_` limitation.

Reply:
Thanks for the detailed reply @brianrourkeboll. I think I got it now. 

Just need to find a way to check for the `'T` constrained call and the explicit  interface call `IParsable.Parse`, does not to seems obvious.

@T-Gro any suggestion for the check ?